### DISCUSSION TITLE
Why useAtom in client component causes hydrate same component twice?

### DISCUSSION THREAD

**Discussion by baeharam** [pmndrs/jotai#2589] [Q&A]
**Test Environment**

- Next.js 14.2.3 app router
- jotai 2.8.2

When I use `useAtom` in client component and setter in rendering process, **it hydrates same component twice.**
Why this kind of issue occurs?

```ts
'use client';

const countAtom = atom(0);

const Test = ({ children }: React.PropsWithChildren) => {
  const [count, setCount] = useAtom(countAtom);
  
  if (count === 0) {
    setCount(1);
  }

  return count === 0 ? null : <>{children}</>
}
```

This client component code inserted into RSC(`layout.tsx` of Next.js). When I run `npm run dev` and refresh several times, it renders 2 component in same page. 

⬇️ Reproduction & Video

|Reproduction Video|
|----|
|<video src="https://github.com/pmndrs/jotai/assets/35518072/53d5d3d3-0984-4573-9c58-84e0c7c2216e">|

**[CodeSandbox ➡️ ](https://codesandbox.io/p/devbox/jotai-atom-twice-rendering-q3mmlx?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clwtdu3l700063b6km3o3akyc%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clwtdu3l700023b6kwlplynua%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clwtdu3l700043b6kw8kt10ih%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clwtdu3l700053b6kdrno24jj%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clwtdu3l700023b6kwlplynua%2522%253A%257B%2522id%2522%253A%2522clwtdu3l700023b6kwlplynua%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clwtdu3l700053b6kdrno24jj%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clwtdu3l700053b6kdrno24jj%2522%257D%252C%2522clwtdu3l700043b6kw8kt10ih%2522%253A%257B%2522id%2522%253A%2522clwtdu3l700043b6kw8kt10ih%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwtdu3l700033b6k4c1qgt0o%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clwtdu3l700033b6k4c1qgt0o%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)**

Did I something wrong?

Thanks for reading!

**Comment by dai-shi** [comment]
I don't know...

But,
```js
  if (count === 0) {
    setCount(1);
  }
```
it may violate rules of React. It's safer to use `useEffect`.
```js
  useEffect(() => {
    if (count === 0) {
      setCount(1);
    }
  })
```

**Reply by baeharam** [reply]
@dai-shi 

I already solved it using by `useEffect`. But, what I wanna know is that why `useAtom` triggers this problem.
`useState` doesn't make this problem... 🤔
